File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/billing/shared/billing-transaction.model.ts
import { BillingTransactionItem } from './billing-transaction-item.model';

import * as moment from 'moment';
import { Visit } from '../../appointments/shared/visit.model';
import { LabTestRequisition } from '../../labs/shared/lab-requisition.model';
import { ImagingItemRequisition } from '../../radiology/shared/imaging-item-requisition.model';
import { ENUM_BillPaymentMode, ENUM_VisitType } from '../../shared/shared-enums';
import { DischargeDetailBillingVM } from '../ip-billing/shared/discharge-bill.view.models';
import { BillingTransactionCreditStatus } from './billing-transaction-credit-status';

export class BillingTransaction {
  public BillingTransactionId: number = 0;
  public PatientId: number = 0;
  public PatientVisitId: number = null;
  //added Patient to retrieve patient information in inter-module call. since patient service gets cleared in such situation.
  //it is already there in server model.//sudarshan:14july
  public Patient: any = null;
  public CounterId: number = 0;
  public PaidDate: string = null;
  public TransactionType: string = null;
  public InvoiceType: string = null;// pratik:29April 2020-- needed for partial payment invoice in ipbilling
  public TotalQuantity: number = 0;
  public SubTotal: number = 0;
  public DiscountPercent: number = 0;
  public DiscountAmount: number = 0;
  public TaxTotal: number = 0;
  public TotalAmount: number = 0;
  public PaidAmount: number = 0;
  public DepositAmount: number = 0;
  public DepositAvailable: number = 0;
  public DepositUsed: number = 0;
  public DepositReturnAmount: number = 0;
  public DepositBalance: number = 0;
  public Remarks: string = null;
  public Tender: number = 0;
  public Change: number = 0;
  //this will be employeeid of the requesting user--sudarshan:7May'17
  public CreatedBy: number = 0;
  public CreatedOn: string = Date();
  public ReturnRemarks: string = null;
  public PrintCount: number = 0;
  public PrintedOn: string = null;
  public PrintedBy: number = null;
  public CreditBalance: number = 0;
  public CreditNoteNumber: number = 0;
  public ReturnedAmount: number = 0;
  public BillingTransactionItems: Array<BillingTransactionItem> = new Array<BillingTransactionItem>();
  public PaymentMode: string = "cash";
  public PaymentDetails: string = null;
  public BillStatus: string = null;
  public FiscalYearId: number = 0;
  public InvoiceNo: number = 0;
  public InvoiceNumber: number = 0; //used while comparing invoiceNummber from //InvoiveNo is already existed.
  public FiscalYear: string = null;
  public TaxId: number = null;
  public InvoiceCode: string = null;
  public TaxableAmount: number = 0;//sud: 9May'18
  public ReturnStatus: boolean = false;
  public IsSelected: boolean = false;//sud: 16may'18 -- Only in client side

  public NonTaxableAmount: number = 0;//added: sud: 29May'18
  public PaymentReceivedBy: number = null;//added: sud: 29May'18
  public PaidCounterId: number = null;//added: sud: 29May'18
  public IsCopyReceipt: boolean = false; //added: Hom: 3rd April 19
  //insurance
  public IsInsuranceBilling: boolean = false;
  public IsInsuranceClaimed: boolean = false;

  public InsuranceClaimedDate: string = null;
  public InsuranceProviderId: number = null;

  public PackageId: number = null;//sud: 10Sept'18-- needs revision
  public PackageName: string = null;//sud: 10Sept'18-- needs revision
  public OrganizationId: number = null;//Yubraj --22nd April '19
  public OrganizationName: string = null;
  public ExchangeRate: number = 0; // Sanjit:5-17-19
  public singleReceiptBool: boolean = false;
  public BillingUserName: string = null; //Yubraj 28th June '19

  public InsTransactionDate: string = null;//sud:19Jul'19--For MNK Insurance Transaction Date..
  public AdjustmentTotalAmount: number = 0;
  public ReturnedItems: any;
  public Ins_NshiNumber: string = '';

  public LabTypeName: string = '';
  public ClaimCode: number = null;//sud:1-oct'21: Changed datatype from String to Number in all places

  public NetAmount: number = 0;//for bill settlement
  public isSelected: boolean = true; //for bill settlement
  public BillReturnIdsCSV: any[] = [];
  public ModuleName: any = null;
  public EmployeeCashTransaction: Array<EmployeeCashTransaction> = new Array<EmployeeCashTransaction>();

  public ReceivedAmount: number = 0; //Krishna,19thAUG'22, Added this property to handle CoPayment.
  public CoPayment_PaymentMode: string = ENUM_BillPaymentMode.credit; //Krishna,19thAUG'22, Added this property to handle CoPayment PaymentMode.
  public IsCoPayment: boolean = false;
  public CoPaymentCreditAmount: number = 0;
  public BillingTxnCreditStatus: Array<BillingTransactionCreditStatus> = new Array<BillingTransactionCreditStatus>();
  public PatientMapPriceCategoryId: number = null;
  public PriceCategoryId: number = null;
  public IsMedicarePatientBilling: boolean = false;
  public SchemeId: number = null;
  public MemberNo: string = null;
  public SalesAmount: number = 0;
  public ReturnAmount: number = 0;
  public InvoiceDate: string = '';
  public InvoiceOf: string = ''; //* Krishna, 24thApril'23, Added this to identify Pharmacy invoice and Billing Invoice in Settlement Page
  public VisitType: string = ENUM_VisitType.outpatient; //Bibek 14thJune'23
  public OtherCurrencyDetail: string = null;
  public IsProvisionalDischargeCleared: boolean = false;
  constructor() {
    this.CreatedOn = moment().format("YYYY-MM-DD HH:mm:ss");
  }
}
export class BillingTransactionPost {
  public LabRequisition: Array<LabTestRequisition> = new Array<LabTestRequisition>();
  public ImagingItemRequisition: Array<ImagingItemRequisition> = new Array<ImagingItemRequisition>();
  public VisitItems: Array<Visit> = new Array<Visit>();
  public Txn: BillingTransaction = new BillingTransaction();
}

export class IPBillTxnVM {
  public billingTransactionModel: BillingTransaction = new BillingTransaction();
  public dischargeDetailVM: DischargeDetailBillingVM = new DischargeDetailBillingVM();
}

export class EmployeeCashTransaction {
  public InAmount: number = 0;
  public OutAmount: number = 0;
  public PaymentModeSubCategoryId: number = 0;
  public ModuleName: any = null;
  public PaymentSubCategoryName: string = '';
  public Remarks: string = null;
  public PaymentDetail: string = null;
}

File: Code/Components/DanpheEMR.ServerModel/BillingModels/BillingTransactionModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DanpheEMR.ServerModel
{
    public class BillingTransactionModel
    {
        [Key]
        public int BillingTransactionId { get; set; }
        public int PatientId { get; set; }
        public int? PatientVisitId { get; set; }
        public int? CounterId { get; set; }
        public DateTime? PaidDate { get; set; }
        public string TransactionType { get; set; }
        public double? TotalQuantity { get; set; }
        public double? SubTotal { get; set; }
        public double? DiscountPercent { get; set; }
        public double? DiscountAmount { get; set; }
        public double? TaxTotal { get; set; }
        public double? TotalAmount { get; set; }
        public double? PaidAmount { get; set; }
        public double? DepositAmount { get; set; }
        public double? DepositAvailable { get; set; }
        public double? DepositUsed { get; set; }
        public double? DepositReturnAmount { get; set; }
        public double? DepositBalance { get; set; }
        public string Remarks { get; set; }
        public double? Tender { get; set; }
        public double? Change { get; set; }

        //this will be employeeid of the requesting user--sudarshan:7May'17
        public int? CreatedBy { get; set; }//might need to create virtual employeemodel for this
        public DateTime? CreatedOn { get; set; }
        //public string BillReturnRemarks { get; set; }//removed sud: 5May'18
        public int? PrintCount { get; set; }

        //added below 3fields: sud:4May'18 - for CreditInvoice scenario
        public string PaymentMode { get; set; }
        public string PaymentDetails { get; set; }
        public string BillStatus { get; set; }
        //start: Ashim-5May for FiscalYear and InvoiceNo.
        public int? FiscalYearId { get; set; }
        [NotMapped]
        public string FiscalYear { get; set; }//added: sud-5May (need to show this in client side)
        public int? InvoiceNo { get; set; }
        //end: Ashim-5May for FiscalYear and InvoiceNo.
        public virtual PatientModel Patient { get; set; }
        public virtual List<BillingTransactionItemModel> BillingTransactionItems { get; set; }

        public bool? ReturnStatus { get; set; }
        //public bool? IsRemoteSynced { get; set; }//added for IRD-sud:6May'18

        public int? TaxId { get; set; }
        public string InvoiceCode { get; set; }
        public double? TaxableAmount { get; set; }//sud:9May'18
        public bool? IsRemoteSynced { get; set; }
        public bool? IsRealtime { get; set; }//sud: 10May'18

        public int? SettlementId { get; set; }//sud: 13May'18--check with dinesh if it's already added in another branch.

        public double? NonTaxableAmount { get; set; }//added: sud: 29May'18
        public int? PaymentReceivedBy { get; set; }//added: sud: 29May'18
        public int? PaidCounterId { get; set; }//added: sud: 29May'18

        public int? PackageId { get; set; }//sud: 10Sept'18-- needs revision
        public string PackageName { get; set; }//sud: 10Sept'18-- needs revision
                                               //15July'2018 :Ashim Added for Govt Insurance Billing
        public bool? IsInsuranceBilling { get; set; }
        public bool? IsInsuranceClaimed { get; set; }
        public DateTime? InsuranceClaimedDate { get; set; }
        public int? InsuranceProviderId { get; set; }
        //Yubraj: 22nd April '19 for credit organization 
        public double? ExchangeRate { get; set; }//Sanjit: 5-17-19 added for foreign exchange
        public int? OrganizationId { get; set; }
        [NotMapped]
        public string OrganizationName { get; set; }

        [NotMapped]//Yubraj 16th Jan '19
        public int? ReceiptNo { get; set; }
        [NotMapped]//shankar 14th nov '19
        public int? CreditNoteNumber { get; set; }
        //sud: 19Jul'19-- For Insurance Transaction Date for MNK-Specific.
        public DateTime? InsTransactionDate { get; set; }
        public DateTime? PrintedOn { get; set; } //Yubraj: 13th August'19
        public int? PrintedBy { get; set; } //Yubraj: 13th August'19

        public int? PartialReturnTxnId { get; set; } // Rajesh:7Aug19
        public decimal? AdjustmentTotalAmount { get; set; }
        [NotMapped]
        public string BillingUserName { get; set; }//Yubaraj:28June'19--needed for Billing receipt to return current logged in user.

        public string InvoiceType { get; set; }// pratik:29April 2020-- needed for partial payment invoice in ipbilling

        public string LabTypeName { get; set; }//pratik:25Feb2021

        //sud:1-Oct'21--Changing Claimcode from String to Int64-- to use Incremental logic (max+1)
        //need nullable since ClaimCode is Non-Mandatory for normal visits.
        public Int64? ClaimCode { get; set; }//pramod

        public static BillingTransactionModel GetCloneWithItems(BillingTransactionModel txnToClone)
        {
            BillingTransactionModel retTxnModel = new BillingTransactionModel()
            {
                BillingTransactionId = txnToClone.BillingTransactionId,
                BillingTransactionItems = txnToClone.BillingTransactionItems,
                FiscalYearId = txnToClone.FiscalYearId,
                InvoiceCode = txnToClone.InvoiceCode,
                InvoiceNo = txnToClone.InvoiceNo,
                SubTotal = txnToClone.SubTotal,
                DiscountAmount = txnToClone.DiscountAmount,
                DiscountPercent = txnToClone.DiscountPercent,
                SettlementId = txnToClone.SettlementId,
                BillStatus = txnToClone.BillStatus,
                PaymentMode = txnToClone.PaymentMode,
                PaymentDetails = txnToClone.PaymentDetails,
                CreatedOn = txnToClone.CreatedOn,
                CreatedBy = txnToClone.CreatedBy,
                CounterId = txnToClone.CounterId,
                DepositAmount = txnToClone.DepositAmount,
                DepositBalance = txnToClone.DepositBalance,
                PatientId = txnToClone.PatientId,
                IsRealtime = txnToClone.IsRealtime,
                IsRemoteSynced = txnToClone.IsRemoteSynced,
                PatientVisitId = txnToClone.PatientVisitId,
                Remarks = txnToClone.Remarks,
                TaxableAmount = txnToClone.TaxableAmount,
                TaxId = txnToClone.TaxId,
                TaxTotal = txnToClone.TaxTotal,
                TransactionType = txnToClone.TransactionType,
                PaidAmount = txnToClone.PaidAmount,
                PaidDate = txnToClone.PaidDate,
                TotalAmount = txnToClone.TotalAmount,
                TotalQuantity = txnToClone.TotalQuantity,
                DepositReturnAmount = txnToClone.DepositReturnAmount,
                Change = txnToClone.Change,
                PrintCount = txnToClone.PrintCount,
                Tender = txnToClone.Tender,
                ReturnStatus = txnToClone.ReturnStatus,
                InvoiceType = txnToClone.InvoiceType,
                LabTypeName = txnToClone.LabTypeName
            };
            return retTxnModel;
        }
    }
    public class BillingTransactionPostVM
    {
        public List<LabRequisitionModel> LabRequisition = new List<LabRequisitionModel>();
        public List<ImagingRequisitionModel> ImagingItemRequisition = new List<ImagingRequisitionModel>();
        public List<VisitModel> VisitItems = new List<VisitModel>();
        public BillingTransactionModel Txn = new BillingTransactionModel();
    }

}
